### ISSUE TITLE
A potential deadlock error due to cyclic order on lock acqusitions

### ISSUE THREAD

**Issue by jenny-cheung** [anlsys/aml#203]
https://github.com/anlsys/aml/blob/5a5fb7edbab59eaa413afd8f342460087765c5c2/src/utils/async.c#L92

```
int aml_active_sched_num_tasks(struct aml_sched *sched)
{
	...;
	// Lock queues to get the right count.
	pthread_mutex_lock(&s->workq_lock);
	pthread_mutex_lock(&s->doneq_lock);
	...;
	pthread_mutex_unlock(&s->workq_lock);
	pthread_mutex_unlock(&s->doneq_lock);
	return n;
}
```

```
// Wait for any task when calling thread is responsible for progress.
struct aml_task *aml_active_sched_wait_any(struct aml_sched_data *data)
{
	...;
	pthread_mutex_lock(&(sched->doneq_lock));
	...;

	pthread_mutex_lock(&(sched->workq_lock));
	...;

	pthread_mutex_unlock(&(sched->workq_lock));
found_in_doneq:
	pthread_mutex_unlock(&(sched->doneq_lock));
	return task;
}
```

**Comment by jenny-cheung** [comment]
Dear developers, thank you for your checking!

This deadlock case is warned by our static code analysis tool.  The functions `aml_active_sched_num_tasks` and `aml_task *aml_active_sched_wait_any` may be executed by different threads. A deadlock may occur resulting from one thread executing the `aml_active_sched_num_tasks` method, acquiring the lock` s->workq_lock` first, and waiting for the lock  `s->doneq_lock` while another thread executing the `aml_task *aml_active_sched_wait_any`, acquiring the lock `sched->doneq_lock` and waiting for the lock` sched->workq_lock`. It is clear that there is a cyclic dependence order on the locks. I think the fix is to adjust the order between these two locks.

Best,

**Comment by jenny-cheung** [comment]
By the way, this case just possibly occurs, because the two functions may not be executed concurrently by two threads. But they are indeed not thread-safe and it is worth fixing because the use of multiple locks should follow a certain order to avoid deadlock. Maybe one day they are involved mistakenly one day that triggers a real deadlock. There is a useful reference about [lock discipline](https://docs.oracle.com/cd/E18752_01/html/816-5137/guide-35930.html) to avoid deadlock, saying "when using multiple locks, avoid deadlocks by making sure that all threads acquire the locks in the same order".

**Comment by NicolasDenoyelle** [comment]
Dear jenny-cheung,
Thank you for opening an issue about this problem.
There was indeed a deadlock in the code you described, as well as in other places in this file.
These have been fixed in the master branch recently as we started using this piece of code in some part of the library. 

We hope not to have left any but there is only so few the human eye can catch. 
Therefore, we will be investigating static analysis tools to include them in our continuous integration and improve the quality of our library.
